jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,459 @@
+            if (maxReservations >= 0 && pending >= maxReservations)
+            {
+                return null;
+            }
+
+            if (!closed && sharedList.size() < maxEntries)
+            {
+                Entry entry = new Entry();
+                sharedList.add(entry);
+                return entry;
+            }
+            return null;
+        }
+        finally
+        {
+            reservationLock.unlock();
+        }
+    }
+
+    /**
+     * Acquire the entry from the pool at the specified index. This method bypasses the thread-local mechanism.
+     *
+     * @param idx the index of the entry to acquire.
+     * @return the specified entry or null if there is none at the specified index or if it is not available.
+     */
+    public Entry acquireAt(int idx)
+    {
+        if (closed)
+            return null;
+
+        try
+        {
+            Entry entry = sharedList.get(idx);
+            if (entry.tryAcquire())
+                return entry;
+        }
+        catch (IndexOutOfBoundsException e)
+        {
+            // no entry at that index
+        }
+        return null;
+    }
+
+    /**
+     * Acquire an entry from the pool.
+     *
+     * @return an entry from the pool or null if none is available.
+     */
+    public Entry acquire()
+    {
+        if (closed)
+            return null;
+
+        // first check the thread-local cache
+        if (cache != null)
+        {
+            List<Entry> cachedList = cache.get();
+            while (!cachedList.isEmpty())
+            {
+                Entry cachedEntry = cachedList.remove(cachedList.size() - 1);
+                if (cachedEntry.tryAcquire())
+                    return cachedEntry;
+            }
+        }
+
+        // then iterate the shared list
+        for (Entry entry : sharedList)
+        {
+            if (entry.tryAcquire())
+                return entry;
+        }
+        return null;
+    }
+
+    /**
+     * This method will return an acquired object to the pool. Objects
+     * that are acquired from the pool but never released will result
+     * in a memory leak.
+     *
+     * @param entry the value to return to the pool
+     * @return false if the pooled entry should be disposed of
+     * @throws NullPointerException if value is null
+     */
+    public boolean release(Entry entry)
+    {
+        if (closed)
+            return false;
+
+        // first mark it as unused
+        boolean reusable = entry.tryRelease();
+
+        // then cache the released entry
+        if (cache != null && reusable)
+        {
+            List<Entry> cachedList = cache.get();
+            if (cachedList.size() < cacheSize)
+                cachedList.add(entry);
+        }
+        return reusable;
+    }
+
+    public boolean isClosed()
+    {
+        return closed;
+    }
+
+    @Override
+    public void close()
+    {
+        reservationLock.lock();
```

Comment:
The only thing that must not happen concurrently is the modification of the shared list, as this could cause:
 - overshooting `maxEntries`
 - overshooting `maxReservations`
 - creating entries in a closed pool

while concurrently acquiring / releasing / closing entries can be safely handled. So this is what this lock is about.

Reply:
As discussed, the lock can be reduced to flip `closed`, copy the entries, and clear `sharedList`. 
The iteration can be outside.